src/primitives/hrp.rs

```diff
@@ -0,0 +1,420 @@
+// SPDX-License-Identifier: MIT
+
+//! Provides an `Hrp` type that represents the human-readable part of a bech32 encoded string.
+//!
+//! > The human-readable part, which is intended to convey the type of data, or anything else that
+//! > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+//! > character having a value in the range [33-126]. HRP validity may be further restricted by
+//! > specific applications.
+//!
+//! ref: [BIP-173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32)
+
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+use core::cmp::Ordering;
+use core::fmt::{self, Write};
+use core::iter::FusedIterator;
+use core::slice::Iter;
+
+use crate::Case;
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug)]
+pub struct Hrp {
+    /// ASCII byte values, guaranteed not to be mixed-case.
+    buf: [u8; MAX_HRP_LEN],
+    /// Number of characters currently stored in this HRP.
+    size: usize,
+}
+
+impl Hrp {
+    /// Parses the human-readable part checking it is valid as defined by [BIP-173].
+    ///
+    /// This does _not_ check that the `hrp` is an in-use HRP within Bitcoin (eg, "bc"), rather it
+    /// checks that the HRP string is valid as per the specification in [BIP-173]:
+    ///
+    /// > The human-readable part, which is intended to convey the type of data, or anything else that
+    /// > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+    /// > character having a value in the range [33-126]. HRP validity may be further restricted by
+    /// > specific applications.
+    ///
+    /// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+    pub fn parse(hrp: &str) -> Result<Self, Error> { Self::parse_and_case(hrp).map(|(p, _)| p) }
+
+    /// Returns the case as well as the parsed `Hrp`.
+    pub(crate) fn parse_and_case(hrp: &str) -> Result<(Self, Case), Error> {
+        use Error::*;
+
+        let mut new = Hrp { buf: [0_u8; MAX_HRP_LEN], size: 0 };
+
+        if hrp.is_empty() {
+            return Err(Empty);
+        }
+        if hrp.len() > MAX_HRP_LEN {
+            return Err(TooLong(hrp.len()));
+        }
+
+        let mut has_lower: bool = false;
+        let mut has_upper: bool = false;
+        for (i, b) in hrp.bytes().enumerate() {
+            // Valid subset of ASCII
+            if !(33..=126).contains(&b) {
+                return Err(InvalidAsciiByte(b));
```

Comment:
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

It would probably be more user-friendly to return a character here instead of a byte.

Reply:
I came to the conclusion that converting a byte that was pulled out of a unicode point was not useful for debugging. This happened when I was debugging some of the current bip test code we have and we were using a unicode string that was actually two bytes and then the first byte caused an error and was printed as a char like you suggest, took me ages to work out what was going on.

Reply:
With the raw invalid byte value at least the dev debugging can think straight away "where did that byte come from" instead of wondering where that weird utf-8 char came from and having to convert it to a byte himself - at least that is what I had to do.

Reply:
So, I think we should be iterating over `hrp.chars()` in this loop, and then if we get a *char* that's out of bounds, we can stick that in the error. After checking it's in bounds we can safely cast it to a u8.

Reply:
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

This is fine but I kinda think we should just fold this into `NonAsciiChar` (and rename this to `InvalidChar`). I don't think the user cares whether an invalid character was in the ASCII range or not.